---
title: "Estrogen receptor binding: data representation and location analysis"
author: "Vince and Rafa"
date: "March 19, 2015"
output: html_document
layout: page
toc: yes
---

```{r options, echo=FALSE}
library(knitr)
opts_chunk$set(fig.path=paste0("figure/", sub("(.*).Rmd","\\1",basename(knitr:::knit_concord$get('infile'))), "-"))
```

```{r setup,echo=FALSE,results="hide",message=FALSE}
library(GenomicFeatures)
library(GenomicRanges)
library(IRanges)
```

# Introduction: Interval data on ER binding

A 2006 Nature Genetics paper (v38, n11) by Carroll, Meyer, Song et al.
describes the notion that a group of genes
whose expression correlates 
with the expression of the estrogen receptor (ER) gene
forms a signature of a breast cancer subtype
called "luminal".  This finding motivates interest in determining
the set of all genomic elements to which ER binds, and this has
been carried out using ChIP-seq experiments.  An important 
finding of the Carroll et al. paper was that ER binding 
in the MCF-7 cell line occurred only rarely promoter-proximal regions.
This study therefore unearthed previously unexplored 
cis-regulatory sites.

We will show how to relate binding peak locations to gene
transcription start sites in two cell lines analyzed in the
ENCODE project.

# Acquisition of binding data

The examples shown here are reported binding sites for estrogen related receptor binding sites derinved from ChIPSeq experiments on two of the main cell lines used by ENCODE: HepG2 and GM12878. In this section we focus on the analysis that comes after the genomic regions are defined. These data can be found as NarrowPeak files on the public repositories but we have created a R package with `GRanges` objects already created. If you have not done so already you can install it like this:

```{r geter,eval=FALSE}
library(devtools)
install_github("genomicsclass/ERBS")
```

Once installed you can attach two object, one for each cell line, and view it's content:
```{r}
library(ERBS)
data(HepG2)  # cell line of liver origin
data(GM12878)  # immortalized B cell
HepG2  # locations of ER binding peaks
```

Note that these object contain two main parts, the regions of the genome which we can extract with  `granges`:
```{r}
granges(HepG2)
```
and other information that we can obtain with either the `mcols` or `values` functions:
```{r}
values(HepG2) ##or mcols(HepG2)
```

<a name="basic"></a>
# Basic operations 

These object are of class `GRanges` defined in the `GenomicsRanges` package:
```{r}
class(HepG2)
```
There are a large number of tools available to use for this class and the best way to learn about them is through the vignette `browseVignettes("GenomicRanges")`

Here we start with some very basic example. Suppose these regions are ordered and we want to work with just the first 10. We can subset these objects as you would expect:
```{r}
granges( HepG2[1:10] )
```
Now what if we want to subset the object by chromosomes? We can access the chromosomes with `seqnames`. Note that not all organisms have chromosomes so Bioconductor uses a more genreal name:

```{r}
chr <- seqnames(HepG2)
chr
```

Note that chr is not a factor or character vector as we would expect. Instead run-length encoding is used. The reason for this is that it provides a much more efficient way to store large granges, for example those obtained from stroing short reads. Note also that we have chromosome names that are not the standard chromosome names such as `chrUn_gl000247`. The human genome actually has some sequences that have not been mapped into one of the chromosomes and are given names like these. 

The Rle class behaves like factors in several useful ways. For example we can tabulate:

```{r}
table(chr)[1:24]
```

We can also subset the object to data based on these. Here is the subset of regions on chr20:

```{r}
granges( HepG2[ chr == "chr20" ] )
```


# Genomic Ranges tools

Before we continue on to finding gene near our binding sites we are going to construct a consensus `GRanges`. Specifically, we are going to construct an object represnenting the regions that are reported for both cell lines. We can easily find these regions using the `findOverlaps` function:

```{r}
res = findOverlaps(HepG2,GM12878)
res
```
 
 TO BE CONTINUED

```{r}
erbs = HepG2[queryHits(res)]
erbs = granges(erbs)
```


# Acquisition of gene transcription start sites

We can use Homo.sapiens:

```{r geths}
library(Homo.sapiens)
ghs = genes(Homo.sapiens)
ghs
```

Explain genes have direction and what is a transcription start site

```{r}
tssgr = resize(ghs,1)
granges(ghs)[1:3,]
granges(tssgr)[1:3,]
```


```{r}
index <- precede(HepG2,ghs)
HepG2[1:3,]
granges(HepG2)[1:3,]
granges(ghs)[index[1:3],]
d = distance(HepG2,ghs[index])
```

If we want the closest to a start site, regardless of befor or after, for example if overlaps. We want something different.

We can create a granges of TSS like this:

```{r}
tssgr = resize(ghs,1)
```


# Finding nearest gene for each binding event

The distanceToNearest function from GenomicFeatures will
create a Hits object with metadata giving the positive distance
between peak location and nearest TSS.  We'll discuss the Hits
object in the video.

```{r donear}
d = distanceToNearest(HepG2, tssgr)
d
##note this gives an error
try(d[,3])
###instead we have to do this
dists = values(d)$distance
```

We will call a distance negative if the peak is 5' to the nearest TSS.
The density estimate given below shows that the vast majority of
events are remote from the 1kb region around TSS; we use dashed
lines to denote that region.

```{r getsi,fig=TRUE}
index = subjectHits(d)
sdists = ifelse(end(HepG2) < start(tssgr[index]), dists, -dists)
hist(sdists, xlim=c(-100000,100000), main="Density of d(ER binding peak, nearest TSS)" ,breaks=seq(min(sdists),max(sdists),len=1000))
abline(v=-c(10000,10000), lty=2)
```

## Getting information about genes

# Annotating genes

Get information on genes

```{r}
index <- subjectHits(d)[dists<1000]
dists = values(d)$distance
##better way to do this?
geneids <- mcols(tssgr[index])$GENEID
```

```{r geths2}
library(Homo.sapiens)
?select
columns(Homo.sapiens)
keytypes(Homo.sapiens)
geneids <- as.character(geneids)
select(Homo.sapiens,keys=geneids,columns="GENENAME",keytype="GENEID")
```






