---
title: "Introduction to Advanced Statistics for the Life Sciences"
author: "Rafa"
date: "January 31, 2015"
output: html_document
layout: page
---

```{r options, echo=FALSE}
library(knitr)
opts_chunk$set(fig.path=paste0("figure/", sub("(.*).Rmd","\\1",basename(knitr:::knit_concord$get('infile'))), "-"))
```

# Installing Bioconductor

Many of the datasets we will use in this chapter require packages made available via the Bioconductor project. Bioconductor is similar to CRAN but uses a different set of functions for downloads. It also includes many more data packages as well as _annotation_ packages that store information about either high-throughout products or information about molecular endpoints such as genes. We will need to some of these packages in this chapter. Here we show how to install the Biobase package. 

```{r,eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("Biobase")
```

You can install a suite of recommended packages by simply typing `biocLite()`

# Data organized in three tables

One of the great advantages of using Bioconductor for high throughput data is that it provides object classes specifically designed to keep high throughput data organized. Below we show an example of how the three tables that are needed to conduct data analysis are available from Bioconductor data objects. For example for gene expression we can use the ExpressionSet object.

```{r,message=FALSE}
library(Biobase)
##can be installed like this: devtools::install_github("genomicsclass/GSE5859")
library(GSE5859)
data(GSE5859)
class(e)
```


These objects were originally designed for gene expression data so the methods to extract the high throughput measurements have related names:
```{r}
dat <- exprs(e)
dim(dat)
```

The information about samples is also stored in this object and the functions to create it try to guarantee that the columns of `exprs(e)` match the rows of the sample information table. `pData` is use as shorthand for _phenotype_ data. 
:

```{r}
sampleInfo <- pData(e)
dim(sampleInfo)
head(sampleInfo)
```

A final table, which we will cover in much more detail in the Bioconductor chapter, is a table that describes the rows, in this case genes. Because each product will have a different table, these have already been created in Bioconductor. Because there are certain products that are widely used, Bioconductor makes databases available from which you can extract this information. This every object does not have to carry around this information:
 
```{r}
library(hgfocus.db)
annot <- select(hgfocus.db, keys=featureNames(e), keytype="PROBEID", columns=c("CHR", "CHRLOC", "SYMBOL"))
##pick one
annot <-annot[match(featureNames(e),annot$PROBEID),]
head(annot)
dim(annot)
```
